### PR TITLE
Change casing of IDs in “Iteration protocol” doc

### PR DESCRIPTION
**Pull request by KennethKinLum**
I saw other places of the MDN docs that the `id` should be all lower cases... and if not changed here, then even the links within this page won't link to them.

### CODE REVIEW

**Comment by sideshowbarker** [comment]
> I saw other places of the MDN docs that the `id` should be all lower cases... and if not changed here, then even the links within this page won't link to them.

Changing the `id` values is not the correct fix; instead, the correct fix is to change all the links on the page to use the titlecase form — for example, `The_iterable_protocol`.

The reason why that’s the correct fix is that even if you manually change the `id` value of an h1-h6 heading in an MDN article, the Yari backend will anyway just auto-generate an `id` value based on the heading text, and Yari will overwrite your manually-set `id` value with that generated value.

And Yari generates titlecased heading IDs in the form, e.g., `The_iterable_protocol` — not all-lowercase.

If you have a local environment to test in, you can try it yourself:

1. Run step 6 in the set of steps at https://github.com/mdn/content/blob/main/README.md#more-substantial-changes.
2. Go to http://localhost:5000/en-US/docs/web/javascript/reference/iteration_protocols in your browser.
3. Use the **Open in your editor** button to edit the source.
4. Change the `id` value of the **The iterable protocol** heading to something like `the_foobarbaz_protocol`.
5. Reload the page.
6. View/inspect the DOM of the page in browser devtools.
   What you‘ll see is that your `the_foobarbaz_protocol` ID value is ignored and instead Yari has automatically given the heading the `id` value `The_iterable_protocol` in the DOM.</div>

**Comment by sideshowbarker** [comment]
> I saw other places of the MDN docs that the `id` should be all lower cases

Whatever places in the MDN docs say that, they’re wrong, as far as I can tell — and we should update them.

**Comment by hamishwillee** [comment]
The docs do not say all other cases, though there has been some recent questions on this.

FWIW though I personally would prefer if yari made all ids lower case, underscore separated by default, and our content followed that. It is then completely predictable and independent of whatever people do with headings now and forevermore.

**Comment by KennethKinLum** [comment]
> The docs do not say all other cases, though there has been some recent questions on this.
> 
> FWIW though I personally would prefer if yari made all ids lower case, underscore separated by default, and our content followed that. It is then completely predictable and independent of whatever people do with headings now and forevermore.

yes, it is just some other places I saw, it is all lower case with underscore.  I agree they should be standardized, or else it is like writing code with Pascal, Ruby, and JS standards all at the same time and it is sometimes snake case and sometimes camel case and getting them to match can waste quite a bit of effort.

**Comment by sideshowbarker** [comment]
> FWIW though I personally would prefer if yari made all ids lower case, underscore separated by default, and our content followed that. It is then completely predictable and independent of whatever people do with headings now and forevermore.

But what value is there in that if Yari just auto-(re)generates `id` values of all headings, and uses those to replace any heading `id` values that you set manually?

Because, as far as I can see, that’s what Yari is doing.

And given that, it’s not clear to me what value there is for us to manual adding `id` values to headings at all, ever.

In fact it seems possibly harmful to manually add `id` values to headings. I say that because the risk is: when we add such a manual `id` value, assuming Yari will actually use it as-is, and we also add links somewhere else in a doc, with href values that have fragments with our manually-set `id` values, then those links are going to break — just as they are broken in the https://developer.mozilla.org/en-US/docs/web/javascript/reference/iteration_protocols article right now.

Or maybe I’m somehow completely wrong about all this. But as far as I can see from local testing at least, I’m not wrong.

**Comment by sideshowbarker** [comment]
> I agree they should be standardized,

It is standardized in practice, as far as I can see: The `id` values in the DOM that Yari generates consistently exactly match the casing of the heading text character-for-character, with the spaces replaced by `_` underscores.

And since basically (with some exceptions) we consistently use a heading style where only the first letter of the first word is uppercase, then basically the `id` values that Yari generates consistently have just the first letter of the first word in uppercase.

> or else it is like writing code with Pascal, Ruby, and JS standards all at the same time and it is sometimes snake case and sometimes camel case and getting them to match can waste quite a bit of effort

Thankfully we don’t actually have that problem — for the reason I outlined above: for `id` values, Yari uses exactly the heading text that you put into the heading, with underscores in place of the spaces.

Again, it’s possible that I may be wrong about what happens on the production site. But at least I can confidently assert that what I’ve described in my comments here is what I’ve observed from testing in my local environment.

In other words, while I can imagine that the behavior of the production site with regard to this is different than the behavior that I see when I test locally, I would be surprised if the behavior of the production site were actually different — because that would mean we can’t trust what we see locally, which would seem… not good.

**Comment by sideshowbarker** [comment]
For the record here, I misspoke in an earlier comment:

> And Yari generates titlecased heading IDs in the form, e.g., `The_iterable_protocol` — not all-lowercase.

That’s not accurate — instead (as I mentioned in a later comment above), Yari generates IDs that exactly match the casing/text of the heading element text content character-for-character (with the spaces replaced by `_` underscores).

**Comment by hamishwillee** [comment]
@sideshowbarker Re https://github.com/mdn/content/pull/1356#issuecomment-762654478 you asked "But what value is there in that if Yari just auto-(re)generates id values of all headings, and uses those to replace any heading id values that you set manually?"

You are almost certainly correct about what Yari does - my point is that what Yari does is "sub-optimal".  What I would prefer it should do is auto-generate a link that is all lowercase, underscore separated, irrespective of the heading.

Why? Back to cognitive load - when I am guessing the anchor I don't have to think about it. Also if you decided that you hated Yari and wanted to use WhateverSSG you would have a cleaner migration path since this is what most anchor generators in markdown do.

So sorry, wasn't discussing "how it is" but what I'd like best. Feel free to ignore.

**Comment by Ryuno-Ki** [comment]
>  lower case, underscore separated

Drive-by comment: This is known as [snake case](https://en.wikipedia.org/wiki/Snake_case)

**Review by sideshowbarker — APPROVED** [review]
